src/PowerShellEditorServices/Session/InvocationEventQueue.cs

```diff
@@ -0,0 +1,251 @@
+                    return;
+                }
+            }
+            finally
+            {
+                _lock.Release();
+            }
+
+            await existingRequest.Task;
+        }
+
+        private async Task SetInvocationRequestAsync(InvocationRequest request)
+        {
+            await WaitForExistingRequestAsync();
+            await _lock.WaitAsync();
+            try
+            {
+                _invocationRequest = request;
+            }
+            finally
+            {
+                _lock.Release();
+            }
+
+            _powerShellContext.ForcePSEventHandling();
+        }
+
+        private void OnPowerShellIdle(object sender, EventArgs e)
+        {
+            if (!_lock.Wait(0))
+            {
+                return;
+            }
+
+            InvocationRequest currentRequest = null;
+            try
+            {
+                if (_invocationRequest == null || System.Console.KeyAvailable)
+                {
+                    return;
+                }
+
+                 currentRequest = _invocationRequest;
+            }
+            finally
+            {
+                _lock.Release();
+            }
+
+            _promptNest.PushPromptContext();
+            try
+            {
+                currentRequest.Invoke(_promptNest.GetPowerShell());
+            }
+            finally
+            {
+                _promptNest.PopPromptContext();
+            }
+        }
+
+        private PSEventSubscriber CreateInvocationSubscriber()
+        {
+            PSEventSubscriber subscriber = _runspace.Events.SubscribeEvent(
+                source: null,
+                eventName: PSEngineEvent.OnIdle,
+                sourceIdentifier: PSEngineEvent.OnIdle,
+                data: null,
+                handlerDelegate: OnPowerShellIdle,
+                supportEvent: true,
+                forwardEvent: false);
+
+            SetSubscriberExecutionThreadWithReflection(subscriber);
+
+            subscriber.Unsubscribed += OnInvokerUnsubscribed;
+
+            return subscriber;
+        }
+
+        private void OnInvokerUnsubscribed(object sender, PSEventUnsubscribedEventArgs e)
+        {
+            CreateInvocationSubscriber();
+        }
+
+        private void SetSubscriberExecutionThreadWithReflection(PSEventSubscriber subscriber)
+        {
+            // We need to create the PowerShell object in the same thread so we can get a nested
+            // PowerShell.  Without changes to PSReadLine directly, this is the only way to achieve
+            // that consistently.  The alternative is to make the subscriber a script block and have
+            // that create and process the PowerShell object, but that puts us in a different
+            // SessionState and is a lot slower.
+
+            // This should be safe as PSReadline should be waiting for pipeline input due to the
+            // OnIdle event sent along with it.
+            typeof(PSEventSubscriber)
```

Comment:
I haven't read closely enough to see if we're using this a lot, but if we are, we should probably cache the `PropertyInfo` so we don't need to perform the reflection call more than once